src/main/java/net/jodah/typetools/TypeResolver.java

```diff
@@ -479,33 +491,58 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
     return genericType instanceof Class ? (Class<?>) genericType : Unknown.class;
   }
 
+  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableTypeMap) {
+    // Check for terminal cases to avoid allocating HashMap and trivial call.
+    if (genericType == null)
+      return null;
+    else if (genericType instanceof Class<?>)
+      return genericType;
+    else
+      return reify(genericType, typeVariableTypeMap, new HashMap<>());
+  }
+
   /**
    * Works like {@link #resolveRawClass(Type, Class, Class)} but does not stop at raw classes. Instead, traverses
    * referenced types.
    */
-  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap) {
-    if (genericType == null) {
+  private static Type reify(Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap, Map<Type, Type> cache) {
+    // Terminal cases.
+    if (genericType == null)
```

Comment:
Will this condition ever be hit?

Reply:
Good question... Will be hard to hit this from a public method. I will try to come up with an example. Pen and paper required...

Reply:
No, this is never being hit. I removed the branch.